File: OpenStreetMap/Shared/DisplayLink.m
//
//  DisplayLink.m
//  Go Map!!
//
//  Created by Bryce on 10/9/14.
//  Copyright (c) 2014 Bryce. All rights reserved.
//

#import "DisplayLink.h"

@implementation DisplayLink

+(instancetype)shared
{
	static DisplayLink * g_shared;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		g_shared = [DisplayLink new];
	});
	return g_shared;
}


-(instancetype)init
{
	self = [super init];
	if ( self ) {
		_blockDict = [NSMutableDictionary new];
		_displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(update:)];
		_displayLink.paused = YES;
		[_displayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
	}
	return self;
}

-(void)update:(CADisplayLink *)displayLink
{
	[_blockDict enumerateKeysAndObjectsUsingBlock:^(NSString * name, void (^block)(void), BOOL *stop) {
		block();
	}];
}

-(double)duration
{
	return _displayLink.duration;
}
-(CFTimeInterval)timestamp
{
	return _displayLink.timestamp;
}

-(void)addName:(NSString *)name block:(void(^)(void))block;
{
	[_blockDict setObject:block forKey:name];
	_displayLink.paused = NO;
}

-(BOOL)hasName:(NSString *)name
{
	return _blockDict[ name ] != nil;
}

-(void)removeName:(NSString *)name;
{
	[_blockDict removeObjectForKey:name];

	if ( _blockDict.count == 0 ) {
		_displayLink.paused = YES;
	}
}

-(void)dealloc
{
	[_displayLink removeFromRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
}
@end

File: src/Shared/DisplayLink.swift
//
//  DisplayLink.swift
//  Go Map!!
//
//  Created by Bryce Cogswell on 10/9/14.
//  Copyright (c) 2014 Bryce Cogswell. All rights reserved.
//

import Foundation
import UIKit

final class DisplayLink {
	static let shared = DisplayLink()

	enum AnimationId {
		case rotateScreenSmoothing // smoothly rotates the screen when compass is enabled/disabled
		case automatedFramerateTest // automatically pans the screen during FPS test
		case fpsLabelUpdate // updates the FPS test label
		case pinDragScroll // pans the screen when pushpin is dragged to edge
		case screenPanningInertia // applies inertia after user pan gesture ends
	}

	var displayLink: CADisplayLink!
	var blockDict: [AnimationId: () -> Void] = [:]

	private static let g_shared = DisplayLink()

	init() {
		displayLink = CADisplayLink(target: self, selector: #selector(step))
		displayLink.isPaused = true
		displayLink.add(to: RunLoop.main, forMode: .default)
		setFrameRate()
	}

	func setFrameRate() {
		if #available(iOS 15.0, *) {
			let maxFPS = Float(UIScreen.main.maximumFramesPerSecond)
			let preferredFPS = (UserPrefs.shared.maximizeFrameRate.value ?? false) ? maxFPS : 60.0
			let rng = CAFrameRateRange(minimum: 30.0,
			                           maximum: maxFPS,
			                           preferred: preferredFPS)
			displayLink.preferredFrameRateRange = rng
		}
	}

	@objc func step() {
		for block in blockDict.values {
			block()
		}
	}

	var duration: CFTimeInterval {
		return displayLink.duration
	}

	var timestamp: CFTimeInterval {
		return displayLink.timestamp
	}

	func add(_ name: AnimationId, block: @escaping () -> Void) {
		blockDict[name] = block
		displayLink.isPaused = false
	}

	func has(_ name: AnimationId) -> Bool {
		return blockDict[name] != nil
	}

	func remove(_ name: AnimationId) {
		blockDict.removeValue(forKey: name)

		if blockDict.count == 0 {
			displayLink.isPaused = true
		}
	}

	deinit {
		displayLink.remove(from: RunLoop.main, forMode: .default)
	}
}
